### PR TITLE
more location encoder refactoring

### LINKED ISSUE #540: location encoder refactoring
I've been thinking lately about `location_encoder.py`, and I've come away with a new appreciation of this code. It's an engineering triumph, solving a really difficult task: inferring a _variable_ number of light sources efficiently on a GPU. Because of this complexity, `location_encoder.py` is also challenging code to understand/maintain/extend. I have some ideas that may help to simplify it.

- [ ] `loc_mean_func` doesn't just apply a sigmoid transformation -- as a special case it returns zero if the input is zero, perhaps to preserve the zeros introduced by the dummy indices and `vpf_padded`? But this is a confusing way to do it, and somewhat brittle, as it relies on floating point equality for control flow. 
- [x] the `loc_mean` parameters are transformed by a sigmoid in `encode_for_n_sources`, whereas other parameters are transformed elsewhere (e.g., `loc_logvar` is transformed to `locs_sd` in `max_a_post`). It'd be better to do all the transformation in one place.
- [x] Perhaps we can delete `_sample_gated_normal` and just directly call `Normal(mean, sd).rsample()`? It seems like the gating isn't helpful frequently, e.g., in the sleep phase training loop. It may be clearer to only do the gating when necessary.
- [x] `var_params` aren't really variational parameters: the variational parameters are the parameters that are fitted by variational inference (i.e., the encoder weights). "Distributional parameters" may be better: it's kind of vague, but at least it isn't incorrect.
- [x] In `max_a_post` don't sample from a normal distribution with zero variance to get the mode, just return the mode/mean directly
- [x] `encode` currently returns all the variational parameter in flat-ish (tile x all_params) tensor that is typically referred to as `var_params_flat`. The thing is, this tensor contains two very different types of distributional parameters. One type governs the number of sources/detections per tile, whereas the other type governs the properties of each possible source/detection. By splitting `var_params_flat` up into two tensors, say `source_params` and `count_params`, `source_params` be reshaped to have dimension (tile x source x params_for_this_source). This way the same parameter for all the sources can be referenced as a slice, e.g., `source_params[:, :, 0:2]` would reference location mean -- no tedious indexing is necessary.
- [x] Don't instantiate an `NllLoss` object every forward pass, just use `F.nll_loss`

### PR DESCRIPTION
**Pull request by jeff-regier**
This PR pretty much closes #540.

I didn't remove some of the masking that should be unnecessary because I was concerned about introducing bugs, and I wasn't sure if I could rely on our tests to catch any new bugs while #563 remains an issue.

There's still quite a bit more we could do to improve the readability of the location encoder. For instance, many of the calls to `rearrange` and `squeeze` could likely be avoided by not flattening structures in the first place. But this PR may be enough for now.

### CODE REVIEW

File: bliss/models/location_encoder.py [hunk 1]
```diff
@@ -216,55 +216,40 @@ def sample(self, var_params: Dict[str, Tensor], n_samples: int) -> Dict[str, Ten
...
+            "fluxes": rearrange(tile_fluxes, unflatten, b=b, nth=nth, ntw=ntw),
+            "n_sources": tile_n_sources,
+        }
 
     def max_a_post(
-        self, var_params: Dict[str, Tensor], n_source_weights: Optional[Tensor] = None
+        self, dist_params: Dict[str, Tensor], n_source_weights: Optional[Tensor] = None
     ) -> TileCatalog:
-        """Derive maximum a posteriori from variational parameters.
+        """Compute the mode of the variational distribution.
```
Comment (jeff-regier) [inline bliss/models/location_encoder.py:249]: It's not exactly a map estimate.

File: bliss/models/location_encoder.py [hunk 2]
```diff
@@ -399,25 +365,20 @@ def training_step(self, batch, batch_idx, optimizer_idx=0):
...
+        }
+        true_catalog = TileCatalog(self.tile_slen, catalog_dict)
 
-        true_catalog = TileCatalog(
-            self.tile_slen,
-            {
-                "locs": batch["locs"][:, :, :, 0 : self.max_detections],
-                "log_fluxes": batch["log_fluxes"][:, :, :, 0 : self.max_detections],
-                "galaxy_bools": batch["galaxy_bools"][:, :, :, 0 : self.max_detections],
-                "n_sources": batch["n_sources"].clamp(max=self.max_detections),
```
Comment (jeff-regier) [inline bliss/models/location_encoder.py:409]: Is all the slicing/clamping necessary here? How would we have more than `self.max_detections` sources in the first place?

File: bliss/models/location_encoder.py [hunk 3]
```diff
@@ -374,16 +335,21 @@ def _encode_for_n_sources(
...
-        var_params_split = torch.split(pps5, split_sizes, 3)
-        names = self.variational_params.keys()
-        var_params_for_n_sources = dict(zip(names, var_params_split))
+        split_sizes = [v["dim"] for v in self.dist_params.values()]
+        dist_params_split = torch.split(pps5, split_sizes, 3)
+        names = self.dist_params.keys()
+        pred = dict(zip(names, dist_params_split))
 
-        # what?!? why is sigmoid(0) = 0?
+        # I don't think the special case for `x == 0` should be necessary
```
Comment (jeff-regier) [inline bliss/models/location_encoder.py:343]: Bryan agrees that this probably isn't necessary. I'm wary of making the change though without more reliable tests

File: bliss/models/location_encoder.py (hunk 2 above)
Comment (dereklhansen) [reply bliss/models/location_encoder.py:409]: For the m2 example, up to five sources can be generated per tile during sleep phase training. We may need to ask Bryan about this

File: bliss/models/location_encoder.py (hunk 1 above)
Comment (dereklhansen) [reply bliss/models/location_encoder.py:249]: I agree - we maybe should rename the ".max_a_post()" methods to be ".mode()" instead, but that can be a separate PR

**Comment by dereklhansen** [comment]
> There's still quite a bit more we could do to improve the readability of the location encoder. For instance, many of the calls to rearrange and squeeze could likely be avoided by not flattening structures in the first place. But this PR may be enough for now.

I agree. I think the only place we absolutely need to flatten is for the conv layer in `.encode()`

**Review by dereklhansen — APPROVED** [review]
Looks great!

File: bliss/models/location_encoder.py (hunk 1 above)
Comment (dereklhansen) [reply bliss/models/location_encoder.py:249]: Opened a new issue
